File: temps/Web-Vue3/src/components/mixin.ts
import FN from '../common/FN'
import { $store } from '../store'
import { IFstate, GlobalObject } from './type.d'
import { calcRect, LAYOUT, px2any, calcUnit } from './client.polyfill'

export default {
  props: {
    hid: {
      type: String,
    },
    clone: {
      type: String,
      required: false,
      default: '',
    },
  },
  computed: {
    APP() {
      return this.$store.app
    },
    SETS() {
      return this.$store.sets
    },
    IT() {
      return this.SETS[this.hid]
    },
    LAYOUT,
    STYLE() {
      if (!this.SETS[this.hid]) return ''

      return this.AP.style
    },
    AP() {
      let hid = this.hid
      let item = this.SETS[hid]
      let clone = this.clone
			let metaName
			let activeStateList: IFstate[] = []
    	let mixinStateList: IFstate[] = []

      item.status.forEach(state => {
				if (!state.active) return
	
				let { name } = state
	
				if (name.includes(':')) return activeStateList.push(state)
	
				if (name == '$mixin') {
					activeStateList.push(state)
	
					if (!metaName) mixinStateList.push(state)
					
					return
				}
	
				if (metaName) return console.warn('meta is repeat', state)
	
				metaName = name
				activeStateList = [...mixinStateList, state]
				mixinStateList = []
			})

      let propsList: IFstate[] = []

      let cloneArr = clone ? clone.split('|').slice(1) : ['0'] // |$|$ => [$, $]

      activeStateList.forEach((subState) => {
				if (subState.name == '$mixin' || !subState.name.includes(':')) {
					return propsList.push(subState)
				}
	
				let nameArr = subState.name.split(':')
				let name = nameArr[0]

				if (name != metaName) return

				let expArr = nameArr.slice(1) // exps => [exp, exp]

				if (!expArr.length) return

				let curr
				let I
				let L = cloneArr.length
				let exp

				for (I = 0; I < L; I++) {
					curr = cloneArr[I]
					exp = expArr[I]

					if (exp) {
						if (!FN.subExpCheck(exp, curr, I, hid)) return
					} else {
						break
					}
				}

				propsList.push(subState)
			})

      let customKeyList: GlobalObject[] = []
      let mixinStyles: GlobalObject[] = []
			let x = 0
			let y = 0
      let tx = 0
    	let ty = 0
			let d = 0
			let s = 100

			propsList.forEach(props => {
				if (props.custom) customKeyList.push(props.custom)

				let { style } = props

				mixinStyles.push(style)

				if (style.x !== undefined) x = style.x
				if (style.y !== undefined) y = style.y
        if (style.tx !== undefined) tx = style.tx
      	if (style.ty !== undefined) ty = style.ty
				if (style.d !== undefined) d = style.d
				if (style.s !== undefined) s = style.s
			})

      let calcProps = propsList[propsList.length - 1] 
			let customKeys = Object.assign({}, ...customKeyList)
			let style = Object.assign({}, ...mixinStyles)
			let mixin = {}

      for (let ckey in customKeys) {
        let ckv = FN.parseModelExp(customKeys[ckey], hid)

        mixin[ckey] =
          typeof ckv == 'string' && ckv.endsWith('px') && !ckv.startsWith('#')
            ? px2any(ckv)
            : ckv
      }

      style.x = x
			style.y = y

      calcRect(style, x, y, tx, ty)

			// The initial value of the zIndex of the static element defaults to 0 if it is not overridden.
			if (style.position == 'static' && style.zIndex === undefined) {
				style.zIndex = 0
			}

			delete style.x
			delete style.y

      style = {
        ...style,
        ...mixin,
      }

      let tfStr = ``

      if (style.perspectValue) {
        tfStr += ` perspective(${calcUnit(Number(style.perspectValue) * 10)})`
  
        if (style.rotateX) tfStr += ` rotateX(${style.rotateX}deg)`
    
        if (style.rotateY) tfStr += ` rotateY(${style.rotateY}deg)`
      }
  
      if (style.scaleX || style.scaleY || style.scaleZ) {
        if (style.scaleX) tfStr += ` scaleX(${Number(style.scaleX) / 100})`
        if (style.scaleY) tfStr += ` scaleY(${Number(style.scaleY) / 100})`
        if (style.scaleZ) tfStr += ` scaleZ(${Number(style.scaleZ) / 100})`
      } else {
        tfStr += ` scale(${s / 100})`
      }
  
      tfStr += ` rotateZ(${d}deg)`
  
      if (style.skewX) tfStr += ` skewX(${style.skewX}deg)`
      if (style.skewY) tfStr += ` skewY(${style.skewY}deg)`
      if (style.translateX) tfStr += ` translateX(${calcUnit(style.translateX)})`
      if (style.translateY) tfStr += ` translateY(${calcUnit(style.translateY)})`
      if (style.translateZ) tfStr += ` translateZ(${calcUnit(style.translateZ)})`
  
      style.transform = tfStr

      let tag = item.model.tag
      
      if (tag && tag.value && this.history.currentTags[tag.value]) {
        FN.PS.publish(hid + 'calcDone', { hid, clone, style, transform: style.transform })

        style.visibility = 'hidden'
        style.transform = ''
      }

      if (this.history.returnTags[hid + clone]) {
        FN.PS.publish(hid + clone + 'calcDone', { hid, clone, style, transform: style.transform })

        style.visibility = 'hidden'
        style.transform = ''
      }

      return {
        style,
        IAA: calcProps.IAA,
        IAD: calcProps.IAD,
      }
    },
    $store() {
      return $store
    },
    history() {
      return this.$store.history
    },
  },
  methods: {
    INIT_MODEL() {},
    GET(key) {
      if (!this.IT) return ''

      let arr = FN.GET_MODEL(this.hid)(key, FN.tfClone(this.clone))
      let v = FN.parseModelExp(Array.isArray(arr) ? arr.toString() : arr, this.hid, false)

      return v
    },
    UPDATE(key, value) {
      FN.SET_MODEL(this.hid)(key, value, FN.tfClone(this.clone))
    },
    CID(hid, ...arg) {
      let item = this.SETS[hid]

      if (!item) return 0

      let cv = item.model.copy.value

      let p = cv

      if (Array.isArray(cv)) {
        ;[...arg].forEach((v) => {
          p = cv[v]
        })

        return Number(p) || 0
      } else {
        return Number(cv) || 0
      }
    },
    CLONE(I) {
      return (this.clone || '') + this.COPY > 0 ? '|' + I : ''
    },
    EV(event, efn, clone) {
      let index = 0

      let hid = efn.name.split('_').slice(1).join('_')

      if (clone && clone.includes('|')) {
        FN.setCurrentClone(hid, clone)

        index = parseInt(clone.split('|').reverse()[0])
      }

      event.context = {
        hid,
        clone,
        index,
        event,
        eventName: event.type,
        response: null,
      }
      event.hid = hid
      event.context.response = efn(event)

      return event.context
    },
    GEV(efn, clone = '') {
      let index = 0

      let hid = efn.name.split('_').slice(1).join('_')
      let useClone = clone && clone.includes('|')

      return (event, payload = {}) => {
        if (useClone) {
          FN.setCurrentClone(hid, clone)

          index = parseInt(clone.split('|').reverse()[0])
        }
        // modelchange, routechange... customEvent
        let eventName

        if (typeof event == 'string') {
          if (event == 'routechange') {
            eventName = 'routechange'
            event = {
              hid,
              value: payload,
            }
          } else {
            let [_, model] = event.split('.')
            eventName = 'modelchange'
            event = {
              hid,
              value: model,
            }
          }

          event.context = {
            hid,
            clone: 0,
            index,
            event,
            eventName,
            response: null,
          }
        } else {
          event.context = {
            hid,
            clone,
            index,
            event,
            eventName: event.gesture ? event.gesture.event[0] : event.type,
            response: null,
          }
          event.hid = hid
        }

        event.context.response = efn(event)
      }
    },
    canRender() {
      let hid = this.hid
      let item = FN.SETS(hid)

      if (!item) return false

      if (!item.model.render) return true

      let render = FN.GET_MODEL(hid)('render', FN.tfClone(this.clone))
      let flag = FN.arrFirst(render)

      if (typeof flag == 'string') {
        flag = FN.parseModelExp(flag, hid, false)
      }

      if (flag === true || flag === 'true') {
        return true
      }

      return false
    },
    hookEnterActive() {
      let ani = this.AP.IAA

      return ani ? ani + 'IA' : ''
    },
    hookLeaveActive() {
      let ani = this.AP.IAD

      return ani ? ani + 'IA' : ''
    },
  },
}

File: temps/flutter/lib/common/mixin.dart
import 'package:flutter/material.dart';
import 'dart:math';
import '../store/index.dart';
import './observer.dart';
import './style.dart';
import './FN.dart';

double unit = 1.0;
double statusBarHeight = 0.0;

final $statusBarState = observe('statusBarState', {}); // rebuild proxy
final baseComponentStyle = {
  'base/level': {'width': 320.0, 'height': 160.0},
  'base/container': {'width': 300.0, 'height': 150.0},
  'base/html': {'width': 250.0, 'height': 150.0},
  'base/photo': {'width': 150.0, 'height': 120.0},
  'base/iframe': {'width': 250.0, 'height': 150.0},
  'base/mirror': {'width': 100.0, 'height': 100.0},
  'base/text': {'width': 80.0, 'height': 22.0},
  'base/link': {'width': 80.0, 'height': 22.0},
  'base/icon': {'width': 45.0, 'height': 45.0},
  'base/input': {'width': 140.0, 'height': 30.0},
  'base/textarea': {'width': 140.0, 'height': 60.0},
  'base/video': {'width': 300.0, 'height': 150.0}
};

void setStatusBarHeight(deviceData) {
  double v = deviceData.padding.top;

  if (v > 0.0) {
    statusBarHeight = v;
  }
}

double deviceWidth = 375.0;
double deviceHeight = 667.0;

/* Use 750 as the design size. */
void setUnit(deviceData) {
  double dw = deviceData.size.width;
  double dh = deviceData.size.height;

  // hack for profile and release mode
  if (dw < 10.0 || dh < 10.0) {
    dw = 375.0;
    dh = 667.0;
  }

  unit = dw / 375.0;
  deviceWidth = dw;
  deviceHeight = dh;

  baseComponentStyle['base/level']!['width'] = deviceWidth / unit;
  baseComponentStyle['base/level']!['height'] = deviceHeight / unit;

  setStatusBarHeight(deviceData);

  $prect['Global'] = {
    'pw': deviceWidth,
    'ph': deviceHeight,
    'pdx': 0.0,
    'pdy': 0.0,
    'bdx': 0.0,
    'bdy': 0.0
  };
}

str2num(str) {
  if (str is int) return str.toDouble();
  if (str is double) return str;
  if (str.endsWith('px')) {
    return double.parse(str.substring(0, str.indexOf('px')));
  }
  return str;
}

double doubleIt(n) {
  if (n is double) return n;
  if (n is int) return n.toDouble();

  return double.tryParse(n) ?? 0.0;
}

double rpx(n) => n * unit;

Map calcAP(hid, clone) {
  var $item = $sets[hid].value;
  var status = $item['status'].value;

  var metaName;
  var activeStateList = [];
  var mixinStateList = [];

  status.forEach((statu) {
    var state = statu.value;

    if (!state['active']) return;

    String name = state['name'];

    if (name.contains(':')) return activeStateList.add(state);

    if (name == '\$mixin') {
      activeStateList.add(state);

      if (metaName == null) mixinStateList.add(state);

      return;
    }

    // ignore: avoid_print
    if (metaName != null) return print('meta is repeat $state');

    metaName = name;
    activeStateList = [...mixinStateList, state];
    mixinStateList = [];
  });

  var propsList = [];
  var cloneArr = clone != '' ? clone.split('|').skip(1).toList() : [ '0' ];

  // ignore: avoid_function_literals_in_foreach_calls
  activeStateList.forEach((subState) {
    if (subState['name'] == '\$mixin' || !subState['name'].contains(':')) {
      return propsList.add(subState);
    }

    List nameArr = subState['name'].split(':');
    String name = nameArr[0];

    if (name != metaName) return;

    List expArr = nameArr.skip(1).toList();

    if (expArr.isNotEmpty) {
      var curr;
      var I;

      int L = cloneArr.length;

      var exp;

      for (I = 0; I < L; I++) {
        curr = cloneArr[I];
        exp = expArr[I];

        if (exp != null) {
          if (!subExpCheck(exp, curr, I, hid)) return;
        } else {
          break;
        }
      }

      propsList.add(subState);
    }
  });

  var customKeyList = [{}];
  var mixinStyles = [{}];
  int x = 0;
  int y = 0;
  int tx = 0;
  int ty = 0;
  int d = 0;

  var s;

  for (var props in propsList) {
    customKeyList.add(props['custom'].value);

    var $style = props['style'].value;

    mixinStyles.add($style);

    if ($style['x'] != null) x = $style['x'];
    if ($style['y'] != null) y = $style['y'];
    if ($style['tx'] != null) tx = $style['tx'];
    if ($style['ty'] != null) ty = $style['ty'];
    if ($style['d'] != null) d = $style['d'];
    if ($style['s'] != null) s = $style['s'];
  }

  Map customKeys = assignObj(customKeyList);
  Map style = assignObj(mixinStyles);
  Map mixinStyle = {};

  customKeys.forEach((K, V) {
    mixinStyle[K] = parseModelExp(V, hid, false);
  });

  return {
    'style': {
      ...style,
      'x': x,
      'y': y,
      'tx': tx,
      'ty': ty,
      'gty': ty,
      'd': d,
      's': s,
      ...mixinStyle
    }
  };
}

final $calcCache = {};

getPosition(hid, clone) {
  if ($calcCache[hid + clone] != null) return $position[hid + clone]; //缓存

  Map ap = calcAP(hid, clone);
  var style = ap['style'];
  var position = style['position'];

  $position[hid + clone] = position;
  $calcCache[hid + clone] = true;

  return position;
}

void numAttr(css, List keys) {
  for (var key in keys) {
    if (css[key] != null) {
      css[key] = rpx(str2num(css[key]));
    }
  }
}

final sideDirection = ['Top', 'Right', 'Bottom', 'Left'];
final fourSide = [0, 1, 2, 3];

List calcSides(css, String type) {
  bool isEmpty = true;
  List padding = [0.0, 0.0, 0.0, 0.0];
  List sides = css[type] == null ? padding : css[type].split(' ').map((e) => rpx(str2num(e))).toList();
  List psides = [];

  for (var i in fourSide) {
    String direct = sideDirection[i];

    if (css[type + direct] == null) {
      psides.add(sides[i]);
    } else {
      psides.add(rpx(str2num(css[type + direct])));

      isEmpty = false;
    }
  }

  return isEmpty && css[type] == null ? [] : psides;
}

Map getStyle(hid, clone) {
  Map item = $struct[hid];

  bool isLevel = item['content'] == 'base/level';

  Map ap = calcAP(hid, clone);
  Map css = ap['style'];
  Map<String, double>? baseRect = baseComponentStyle[item['content']];

  css['x'] = rpx(css['x']);
  css['y'] = rpx(css['y']);
  css['rotate'] = css['d'];

  var w = css['width'] == null ? baseRect!['width'] : str2num(css['width']);
  var h = css['height'] == null ? baseRect!['height'] : str2num(css['height']);

  Map prect = $prect[hid + clone] ?? $prect['Global'] ;

  bool isStatic = css['position'] == 'static';

  if (w is String && w.endsWith('%')) {
    double dx = isStatic ? (prect['pdx'] + prect['bdx']) : prect['bdx'];

    css['width'] = (prect['pw'] - dx) / 100.0 * double.parse(w.substring(0, w.indexOf('%')));
  } else {
    css['width'] = rpx(w);
  }

  if (h is String && h.endsWith('%')) {
    double dy = isStatic ? (prect['pdy'] + prect['bdy']) : prect['bdy'];

    css['height'] = (prect['ph'] - dy) / 100.0 * double.parse(h.substring(0, h.indexOf('%')));
  } else {
    css['height'] = rpx(h);
  }

  css['boxShadows'] = css['boxShadow'] == null
  ? []
  : css['boxShadow'].split('inset, ').map((v) => v += v.contains('inset') ? '' : 'inset').toList();

  if (isLevel && css['useSafeArea']) {
    if ($statusBarState.value[hid] == false) {
      css['height'] -= statusBarHeight;
      css['y'] += statusBarHeight;
    }
  }

  if ($safePosition[hid] == true && css['ty'] == 0) {
    css['y'] += statusBarHeight;
  }

  return css;
}

Map calcStyle(hid, clone) {
  var css = getStyle(hid, clone);

  css['rectWidth'] = css['width'];
  css['rectHeight'] = css['height'];
  css['nx'] = 0.0;
  css['ny'] = 0.0;

  List psides = calcSides(css, 'padding');

  if (psides.length > 1) {
    css['padding'] = EdgeInsets.only(top: psides[0], right: psides[1], bottom: psides[2], left: psides[3]);
    css['pdx'] = psides[3] + psides[1];
    css['pdy'] = psides[0] + psides[2];
    css['paddingSide'] = psides;
  } else {
    css['paddingSide'] = [0.0, 0.0, 0.0, 0.0];
  }

  if (css['borderWidth'] != null) {
    var sides = css['borderWidth'].split(' ');

    if (sides.length > 1) {
      sides = sides.map((e) => rpx(str2num(e))).toList();

      css['borderWidths'] = sides;
      css['width'] = css['width'] - sides[1] - sides[3];
      css['height'] = css['height'] - sides[0] - sides[2];
      css['bdx'] = sides[1] + sides[3];
      css['bdy'] = sides[0] + sides[2];
    } else {
      var sideWidth = rpx(str2num(sides[0]));

      css['borderWidths'] = [sideWidth, sideWidth, sideWidth, sideWidth];

      css['bdx'] = css['bdy'] = sideWidth * 2.0;
    }
  }

  if (css['borderRadius'] != null) {
    var sides =
        css['borderRadius'].split(' ').map((e) => rpx(str2num(e)));
    List borderSides = sides.map((v) => Radius.circular(v)).toList();
    css['borderRadiusValue'] = sides.toList();
    css['borderRadius'] = BorderRadius.only(
      topLeft: borderSides[0],
      topRight: borderSides[1],
      bottomRight: borderSides[2],
      bottomLeft: borderSides[3],
    );
  } else {
    css['borderRadiusValue'] = [0.0, 0.0, 0.0, 0.0];
  }

  css['marginSize'] = [0.0, 0.0];

  List msides = calcSides(css, 'margin');

  if (msides.length > 1) {
    css['margin'] = EdgeInsets.only(top: msides[0], right: msides[1], bottom: msides[2], left: msides[3]);
    css['marginSize'] = [msides[3], msides[0]]; //[left, top]
  }

  if (css['backgroundColor'] != null) {
    css['backgroundColor'] = tfColor(css['backgroundColor']);
  }

  if (css['color'] != null) {
    css['color'] = tfColor(css['color']);
  }

  if (css['fill'] != null) {
    css['fill'] = tfColor(css['fill']);
  }

  if (css['borderColor'] != null) {
     css['borderColor'] = tfColor(css['borderColor']);
  }

  if (css['filter'] != null) {
    css['filter'] = tfFilter(css['filter']);
  }

  if (css['backdropFilter'] != null) {
    css['backdropFilter'] = tfFilter(css['backdropFilter']);
  }

  numAttr(css, ['fontSize', 'lineHeight', 'letterSpacing']);

  double k = pi / 180;

  css['perspectValue'] = doubleIt(css['perspectValue'] ?? 0.0) * 10;
  css['rotateX'] = doubleIt(css['rotateX'] ?? 0.0) * k;
  css['rotateY'] = doubleIt(css['rotateY'] ?? 0.0) * k;
  css['rotateZ'] = doubleIt(css['rotateZ'] ?? 0.0) * k;
  css['skewX'] = doubleIt(css['skewX'] ?? 0.0) * k;
  css['skewY'] = doubleIt(css['skewY'] ?? 0.0) * k;
  css['translateX'] = doubleIt(css['translateX'] ?? 0.0) * unit;
  css['translateY'] = doubleIt(css['translateY'] ?? 0.0) * unit;
  css['translateZ'] = doubleIt(css['translateZ'] ?? 0.0) * unit;

  if (css['transition'] != null) {
    var tarr = parseTransition(css['transition']);

    css['during'] = tarr[0] * 1000.0; //s to ms
    css['curve'] = tarr[1];
  }

  return css;
}

const $padding = SizedBox.shrink();
const $zeroEdge = EdgeInsets.only(top: 0.0, right: 0.0, bottom: 0.0, left: 0.0);

List generateArray(n) {
  return List.filled(n.toInt(), 0, growable: true);
}

String setItem(id) {
  return id;
}

int getCloneNum(id, [clone = '']) {
  var copy = GET_(id, '', 'copy');

  if (copy == null) return 1;
  if (copy is int) return copy;
  if (copy is double) return copy.round();

  var mcopy = FN.GET_MODEL(id)('copy', tfClone(clone));
  var flag = arrFirst(mcopy);

  if (flag is String && flag.substring(0, 1) == '\$') {
    flag = parseModelExp(flag, id, false);
  }

  return int.tryParse(flag) ?? 0;
}
getSubClone(id, int J, [clone = '']) {
  var copy = GET_(id, clone, 'copy');

  bool ishasCn = copy != null;
  bool isParentHasCn = clone.contains('|');

  if (isParentHasCn) {
    if (ishasCn) {
      return clone + '|' + J.toString();
    } else {
      return clone;
    }
  } else {
    if (ishasCn) {
      return '|' + J.toString();
    } else {
      return '';
    }
  }
}

bool canRender(id) {
  var render = GET_(id, '', 'render');

  if (render == null) return true;
  if (render == 'false') return false;
  if (render is bool) return render;

  return true;
}

Map calcState(hid) {
  var item = $struct[hid];
  var $item = $sets[hid].value;
  var config = {
    'hid': setItem(hid),
    'item': $item,
    'clone': '',
    'type': item['type'],
    'style': calcStyle(hid, '')
  };
  return config;
}

class RenderPage extends StatelessWidget {
  final child;
  RenderPage(this.child);
  @override
  Widget build(BuildContext context) {
    return child;
  }
}

Widget transition(slot) {
  return slot;
}

final FractionalTowardMap = {
  'left': 0.0,
  'center': 0.5,
  'right': 1.0,
  'top': 0.0,
  'bottom': 1.0
};

FractionalOffset parseTransformOrigin(String input) {
  if (input == 'center') return FractionalOffset.center;

  List<String> arr = input.split(' ');

  String a = arr[0];
  String b = arr[1];

  return FractionalOffset(
    FractionalTowardMap[a] ?? ((double.tryParse(a.substring(0, a.length - 1)) ?? 0.0) / 100.0),
    FractionalTowardMap[b] ?? ((double.tryParse(b.substring(0, b.length - 1)) ?? 0.0) / 100.0)
  );
}

double amendValue(double a, double b) {
  return sqrt(a * a + b * b) / 2.0 * -1;
}
